envoy/event/deferred_deletable.h

```diff
@@ -13,6 +13,12 @@ namespace Event {
 class DeferredDeletable {
 public:
   virtual ~DeferredDeletable() = default;
+
+  /**
+   * Called when an object is passed to `deferredDelete`. This signals that the object will soon
+   * be deleted.
+   */
+  virtual void deleteIsPending() {}
```

Comment:
@mattklein123 let me know what you think of this. I used it to add more asserts that nobody uses the conn pool after it is removed from the cluster manager; I thought the idea was generally useful.

Reply:
Sure makes sense to me. We could probably do this elsewhere also.